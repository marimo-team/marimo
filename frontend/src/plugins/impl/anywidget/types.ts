/* Copyright 2024 Marimo. All rights reserved. */
/* eslint-disable @typescript-eslint/no-invalid-void-type */
/* eslint-disable @typescript-eslint/no-explicit-any */

// Copied from https://github.com/manzt/anywidget/blob/main/packages/types/index.ts
// with slight modifications
// - removed widget_manager
// We copy this for now since `@anywidget/types` pulls in jquery and backbone, which we don't need

type Awaitable<T> = T | Promise<T>;
export type EventHandler = (...args: any[]) => void;
type ObjectHash = Record<string, any>;
export type ChangeEventHandler<Payload> = (_: unknown, value: Payload) => void;
type LiteralUnion<T, U = string> = T | (U & {});

export interface AnyModel<T extends ObjectHash = ObjectHash> {
  get<K extends keyof T>(key: K): T[K];
  set<K extends keyof T>(key: K, value: T[K]): void;
  off<K extends keyof T>(
    eventName?: LiteralUnion<`change:${K & string}` | "msg:custom"> | null,
    callback?: EventHandler | null,
  ): void;
  on(
    eventName: "msg:custom",
    callback: (msg: any, buffers: DataView[]) => void,
  ): void;
  on<K extends `change:${keyof T & string}`>(
    eventName: K,
    callback: K extends `change:${infer Key}`
      ? ChangeEventHandler<T[Key]>
      : never,
  ): void;
  on(eventName: string, callback: EventHandler): void;
  save_changes(): void;
  send(
    content: any,
    callbacks?: any,
    buffers?: ArrayBuffer[] | ArrayBufferView[],
  ): void;
  widget_manager: {};
}

export interface Experimental {
  invoke: <T>(
    name: string,
    msg?: any,
    options?: {
      buffers?: DataView[];
      signal?: AbortSignal;
    },
  ) => Promise<[T, DataView[]]>;
}
export interface RenderProps<T extends ObjectHash = ObjectHash> {
  model: AnyModel<T>;
  el: HTMLElement;
  experimental: Experimental;
}
export type Render<T extends ObjectHash = ObjectHash> = (
  props: RenderProps<T>,
) => Awaitable<void | (() => Awaitable<void>)>;
export interface InitializeProps<T extends ObjectHash = ObjectHash> {
  model: AnyModel<T>;
  experimental: Experimental;
}

export type Initialize<T extends ObjectHash = ObjectHash> = (
  props: InitializeProps<T>,
) => Awaitable<void | (() => Awaitable<void>)>;
interface WidgetDef<T extends ObjectHash = ObjectHash> {
  initialize?: Initialize<T>;
  render?: Render<T>;
}
export type AnyWidget<T extends ObjectHash = ObjectHash> =
  | WidgetDef<T>
  | (() => Awaitable<WidgetDef<T>>);
