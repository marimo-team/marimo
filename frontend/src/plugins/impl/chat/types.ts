/* Copyright 2024 Marimo. All rights reserved. */
export type ChatRole = "system" | "user" | "assistant";

export interface ChatMessage {
  role: ChatRole;
  content: string;
  attachments?: ChatAttachment[];
}

export interface ChatAttachment {
  name?: string;
  contentType?: string;
  url: string;
}

export interface SendMessageRequest {
  messages: ChatMessage[];
  config: {
    max_tokens?: number;
    temperature?: number;
    top_p?: number;
    top_k?: number;
    frequency_penalty?: number;
    presence_penalty?: number;
  };
}

export interface ChatConfig {
  maxTokens: number;
  temperature: number;
  topP: number;
  topK: number;
  frequencyPenalty: number;
  presencePenalty: number;
}
