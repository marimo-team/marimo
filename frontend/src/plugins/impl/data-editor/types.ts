/* Copyright 2024 Marimo. All rights reserved. */

import type { GridCellKind, GridColumn } from "@glideapps/glide-data-grid";
import type { FieldTypesWithExternalType } from "@/components/data-table/types";
import type { DataType } from "@/core/kernel/messages";

export interface PositionalEdit {
  rowIdx: number;
  columnId: string;
  value: unknown;
}

export enum BulkEdit {
  Insert = "insert",
  Remove = "remove",
  Rename = "rename",
}

export interface RowEdit {
  rowIdx: number;
  type: BulkEdit;
}

export interface ColumnEdit {
  columnIdx: number;
  newName?: string;
  type: BulkEdit;
}

export interface Edits {
  edits: Array<PositionalEdit | RowEdit | ColumnEdit>;
}

export type ModifiedGridColumn = GridColumn & {
  kind: GridCellKind;
  dataType: DataType;
};

export interface DataEditorProps<T> {
  data: T[];
  fieldTypes: FieldTypesWithExternalType | null | undefined;
  onAddEdits: (edits: Edits["edits"]) => void;
  onAddRows: (newRows: object[]) => void;
  columnSizingMode: "fit" | "auto";
}
