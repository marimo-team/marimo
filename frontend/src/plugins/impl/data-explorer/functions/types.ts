/* Copyright 2024 Marimo. All rights reserved. */

import { type AggregateOp, isString } from "vega";
import type {
  ArgmaxDef,
  ArgminDef,
} from "vega-lite/types_unstable/aggregate.d.ts";
import type { isAggregateOp as isAggregateOpVega } from "vega-lite/types_unstable/aggregate.js";

// Vega doesn't expose the constant, so we define all here
const AGGREGATE_OPS: readonly AggregateOp[] = [
  "argmax",
  "argmin",
  "average",
  "count",
  "distinct",
  "exponential",
  "exponentialb",
  "product",
  "max",
  "mean",
  "median",
  "min",
  "missing",
  "q1",
  "q3",
  "ci0",
  "ci1",
  "stderr",
  "stdev",
  "stdevp",
  "sum",
  "valid",
  "values",
  "variance",
  "variancep",
] as const;

// Subset of aggregate operations that we support
const SUPPORTED_AGGREGATE_OPS: readonly AggregateOp[] = [
  "average",
  "count",
  "distinct",
  "max",
  "mean",
  "median",
  "min",
  "q1",
  "q3",
  "stderr",
  "stdev",
  "sum",
];
type SupportedAggregateOp = (typeof SUPPORTED_AGGREGATE_OPS)[number];

// We implement our own isAggregateOp because
// The vega-lite types_unstable import path fails in Vite (module resolution issue)
export const isAggregateOp: typeof isAggregateOpVega = (
  a: string | ArgminDef | ArgmaxDef,
): a is AggregateOp => {
  if (!isString(a)) {
    return false;
  }

  return AGGREGATE_OPS.includes(a as AggregateOp);
};

// Subset of time units that we support
export type TimeUnitOp =
  | "year"
  | "month"
  | "date"
  | "day"
  | "hours"
  | "minutes"
  | "seconds"
  | "milliseconds"
  | "yearmonth"
  | "yearmonthdate"
  | "monthdate";

export type FieldFunction = SupportedAggregateOp | "bin" | TimeUnitOp;
