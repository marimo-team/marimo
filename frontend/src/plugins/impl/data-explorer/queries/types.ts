/* Copyright 2024 Marimo. All rights reserved. */
import { Query } from "compassql/build/src/query";
import { VisualizationSpec } from "react-vega";
import { FieldDefinition, EncodingChannel } from "../encoding";
import { NamedData } from "vega-lite/build/src/data";
import { TopLevel, FacetedUnitSpec } from "vega-lite/build/src/spec";

export interface PlotFieldInfo {
  fieldDef: FieldDefinition;
  channel: EncodingChannel;
}

export interface ResultPlot {
  fieldInfos: PlotFieldInfo[];

  /**
   * Spec to be used for rendering.
   */
  spec: VisualizationSpec;
}

export interface Result {
  plots: ResultPlot[] | null;

  query: Query;

  limit: number;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type TopLevelFacetedUnitSpec = TopLevel<FacetedUnitSpec<any, any>> & {
  data: NamedData;
};

export interface ResultingCharts {
  main: Result;

  histograms: Result;

  addCategoricalField: Result;
  addQuantitativeField: Result;
  addTemporalField: Result;

  alternativeEncodings: Result;
  summaries: Result;
}

export interface QueryCreator {
  type: keyof ResultingCharts;

  limit: number;

  createQuery(query: Query): Query;
}
