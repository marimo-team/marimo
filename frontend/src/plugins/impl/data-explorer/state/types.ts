/* Copyright 2024 Marimo. All rights reserved. */
import type { Schema } from "compassql/build/src/schema";
import type { SpecificEncoding } from "../encoding";
import type { SpecQuery } from "compassql/build/src/query/spec";
import type { SpecMark } from "../marks";

export interface ChartSpec {
  /**
   * The type of the mark.
   */
  mark: SpecMark;

  /**
   * Mapping between specific encoding channels and encoding definitions.
   */
  encoding: SpecificEncoding;

  /**
   * Configuration of the chart.
   */
  config: SpecQuery["config"];

  /**
   * The schema for the data.
   */
  schema?: Schema | null;
}
