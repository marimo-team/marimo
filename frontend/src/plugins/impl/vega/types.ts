/* Copyright 2024 Marimo. All rights reserved. */
import type {
  TopLevelUnitSpec,
  GenericUnitSpec,
} from "vega-lite/build/src/spec/unit";
import type { Encoding } from "vega-lite/build/src/encoding";
import type { Field } from "vega-lite/build/src/channeldef";

export type { SharedCompositeEncoding } from "vega-lite/build/src/compositemark";
export type { AnyMark, MarkDef } from "vega-lite/build/src/mark";
export { Mark } from "vega-lite/build/src/mark";
export type { TopLevelSpec as VegaLiteSpec } from "vega-lite";
export { type Field } from "vega-lite/build/src/channeldef";
export type { SingleDefUnitChannel } from "vega-lite/build/src/channel";
export type {
  SelectionParameter,
  SelectionType,
} from "vega-lite/build/src/selection";
export type { DataFormat } from "vega-lite/build/src/data";
export type {
  LayerSpec,
  UnitSpec,
  GenericFacetSpec,
  FacetedUnitSpec,
} from "vega-lite/build/src/spec";

export type VegaLiteUnitSpec = TopLevelUnitSpec<Field>;
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type GenericVegaSpec = GenericUnitSpec<any, any, any>;
export type EncodingType = keyof Encoding<Field>;
export type Encodings = Encoding<Field>;
