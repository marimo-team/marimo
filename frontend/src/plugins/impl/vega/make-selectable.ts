/* Copyright 2024 Marimo. All rights reserved. */
import {
  GenericVegaSpec,
  Mark,
  SelectionType,
  VegaLiteSpec,
  VegaLiteUnitSpec,
} from "./types";
import { findEncodedFields, makeEncodingInteractive } from "./encodings";
import { Params } from "./params";
import { Marks } from "./marks";

export function makeSelectable<T extends VegaLiteSpec>(
  spec: T,
  opts: {
    chartSelection?: boolean | "interval" | "point";
    fieldSelection?: boolean | string[];
  },
): T {
  // Both default to true
  const { chartSelection = true, fieldSelection = true } = opts;

  // Disable selection if both are false
  if (!chartSelection && !fieldSelection) {
    return spec;
  }

  if ("vconcat" in spec) {
    const subSpecs = spec.vconcat.map((subSpec) =>
      "mark" in subSpec ? makeChartInteractive(subSpec) : subSpec,
    );
    return { ...spec, vconcat: subSpecs };
  }

  if ("hconcat" in spec) {
    const subSpecs = spec.hconcat.map((subSpec) =>
      "mark" in subSpec ? makeChartInteractive(subSpec) : subSpec,
    );
    return { ...spec, hconcat: subSpecs };
  }

  if ("layer" in spec) {
    const subSpecs = spec.layer.map((subSpec, idx) => {
      if (!("mark" in subSpec)) {
        return subSpec;
      }
      let resolvedSpec = subSpec as VegaLiteUnitSpec;
      resolvedSpec = makeChartSelectable(resolvedSpec, chartSelection, idx);
      resolvedSpec = makeChartInteractive(resolvedSpec);
      return resolvedSpec;
    });
    return { ...spec, layer: subSpecs };
  }

  if (!("mark" in spec)) {
    return spec;
  }

  let resolvedSpec: VegaLiteUnitSpec = spec;
  resolvedSpec = makeLegendSelectable(resolvedSpec, fieldSelection);
  resolvedSpec = makeChartSelectable(resolvedSpec, chartSelection, undefined);
  resolvedSpec = makeChartInteractive(resolvedSpec);

  return resolvedSpec as T;
}

/**
 * Given a spec, add the necessary parameters to make the legend selectable.
 */
function makeLegendSelectable(
  spec: VegaLiteUnitSpec,
  fieldSelection: boolean | string[],
): VegaLiteUnitSpec {
  // If fieldSelection is false, we don't do anything
  if (fieldSelection === false) {
    return spec;
  }

  let legendFields = findEncodedFields(spec);
  // If fieldSelection is an array, we filter the fields
  if (Array.isArray(fieldSelection)) {
    legendFields = legendFields.filter((field) =>
      fieldSelection.includes(field),
    );
  }

  const legendParams = legendFields.map((field) => Params.legend(field));
  const nextParams = [...(spec.params || []), ...legendParams];

  return {
    ...spec,
    params: nextParams,
  } as VegaLiteUnitSpec;
}

/**
 * Given a spec, add the necessary parameters to make the chart selectable.
 */
function makeChartSelectable(
  spec: VegaLiteUnitSpec,
  chartSelection: boolean | "interval" | "point",
  /**
   * If the spec is part of a layer, we need to know the layer number.
   * This is so we can give unique names to the parameters.
   */
  layerNum: number | undefined,
): VegaLiteUnitSpec {
  // If chartSelection is false, we don't do anything
  if (chartSelection === false) {
    return spec;
  }

  let mark: Mark;
  try {
    mark = Marks.getMarkType(spec.mark);
  } catch {
    return spec;
  }

  // We don't do anything if the mark is text
  if (mark === "text") {
    return spec;
  }

  const resolvedChartSelection =
    chartSelection === true ? getBestSelectionForMark(mark) : [chartSelection];

  if (!resolvedChartSelection) {
    return spec;
  }

  const params = resolvedChartSelection.map((selectionType) =>
    selectionType === "interval"
      ? Params.interval(spec, layerNum)
      : Params.point(spec, layerNum),
  );

  const nextParams = [...(spec.params || []), ...params];

  return {
    ...spec,
    params: nextParams,
  } as VegaLiteUnitSpec;
}

/**
 * Makes a chart clickable and adds an opacity encoding to the chart.
 */
function makeChartInteractive<T extends GenericVegaSpec>(spec: T): T {
  const prevEncodings = "encoding" in spec ? spec.encoding : undefined;
  const params = spec.params || [];
  const paramNames = params.map((param) => param.name);

  if (params.length === 0) {
    return spec;
  }

  const mark = Marks.getMarkType(spec.mark);
  // We don't do anything if the mark is text
  if (mark === "text") {
    return spec;
  }

  return {
    ...spec,
    mark: Marks.makeClickable(spec.mark),
    encoding: makeEncodingInteractive(
      "opacity",
      prevEncodings || {},
      paramNames,
      spec.mark,
    ),
  };
}

function getBestSelectionForMark(mark: Mark): SelectionType[] | undefined {
  switch (mark) {
    case "text":
    case "arc":
    case "area":
      return ["point"];
    case "bar":
      return ["point", "interval"];
    // there is no best selection for line
    case "line":
      return undefined;
    default:
      return ["point", "interval"];
  }
}
