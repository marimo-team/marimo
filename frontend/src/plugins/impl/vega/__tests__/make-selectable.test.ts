/* Copyright 2024 Marimo. All rights reserved. */
import { expect, describe, it } from "vitest";
import { makeSelectable } from "../make-selectable";
import { VegaLiteSpec } from "../types";

describe("makeSelectable", () => {
  it("should return correctly if mark is not string", () => {
    const spec = {
      mark: {
        type: "point",
      },
    } as VegaLiteSpec;
    expect(makeSelectable(spec, {})).toMatchSnapshot();
  });

  it("should return correctly if mark is a string", () => {
    const spec = {
      mark: "point",
    } as VegaLiteSpec;
    expect(makeSelectable(spec, {})).toMatchSnapshot();
  });

  it("should return the same spec if selection is false", () => {
    const spec = {
      mark: "point",
    } as VegaLiteSpec;
    expect(
      makeSelectable(spec, {
        chartSelection: false,
        fieldSelection: false,
      }),
    ).toEqual(spec);
  });

  it("should return the same spec for not-defined and true", () => {
    const spec = {
      mark: "point",
    } as VegaLiteSpec;
    expect(
      makeSelectable(spec, {
        chartSelection: true,
        fieldSelection: true,
      }),
    ).toEqual(makeSelectable(spec, {}));
  });

  it("should return the same spec if mark is not in spec", () => {
    const spec1 = {
      mark: "point",
    } as VegaLiteSpec;
    const spec2 = {
      mark: {
        type: "point",
      },
    } as VegaLiteSpec;
    expect(makeSelectable(spec1, {})).toEqual(makeSelectable(spec2, {}));
  });

  it("should return correctly if overlapping encodings", () => {
    const spec = {
      config: {
        view: {
          continuousHeight: 300,
        },
      },
      data: { url: "data/cars.json" },
      encoding: {
        color: {
          field: "Origin",
          type: "nominal",
        },
        x: {
          field: "Horsepower",
          type: "quantitative",
        },
        y: {
          field: "Miles_per_Gallon",
          type: "quantitative",
        },
      },
      mark: {
        type: "point",
      },
    } as VegaLiteSpec;
    expect(
      makeSelectable(spec, {
        chartSelection: true,
        fieldSelection: true,
      }),
    ).toMatchSnapshot();
  });

  it("should skip field selection if empty or false", () => {
    const spec = {
      config: {
        view: {
          continuousHeight: 300,
        },
      },
      data: { url: "data/cars.json" },
      encoding: {
        color: {
          field: "Origin",
          type: "nominal",
        },
        x: {
          field: "Horsepower",
          type: "quantitative",
        },
        y: {
          field: "Miles_per_Gallon",
          type: "quantitative",
        },
      },
      mark: {
        type: "point",
      },
    } as VegaLiteSpec;

    expect(
      makeSelectable(spec, { chartSelection: true, fieldSelection: false }),
    ).toMatchSnapshot();

    // These are the same
    expect(
      makeSelectable(spec, { chartSelection: true, fieldSelection: false }),
    ).toEqual(
      makeSelectable(spec, { chartSelection: true, fieldSelection: [] }),
    );
  });

  it("should return correctly with multiple encodings", () => {
    const spec = {
      mark: "point",
      encoding: {
        color: {
          field: "colorField",
          type: "nominal",
        },
        size: {
          field: "sizeField",
          type: "quantitative",
        },
      },
    } as VegaLiteSpec;

    expect(makeSelectable(spec, {})).toMatchSnapshot();
  });

  it("should return correctly if existing legend selection", () => {
    const spec = {
      config: {
        view: {
          continuousHeight: 300,
        },
      },
      data: {
        url: "https://cdn.jsdelivr.net/npm/vega-datasets@v1.29.0/data/unemployment-across-industries.json",
      },
      encoding: {
        color: {
          field: "series",
          scale: { scheme: "category20b" },
          type: "nominal",
        },
        opacity: {
          condition: { param: "param_1", value: 1 },
          value: 0.2,
        },
        x: {
          axis: { domain: false, format: "%Y", tickSize: 0 },
          field: "date",
          timeUnit: "yearmonth",
          type: "temporal",
        },
        y: {
          aggregate: "sum",
          axis: null,
          field: "count",
          stack: "center",
          type: "quantitative",
        },
      },
      mark: { type: "area" },
      params: [
        {
          bind: "legend",
          name: "param_1",
          select: { fields: ["series"], type: "point" },
        },
      ],
    } as VegaLiteSpec;
    expect(makeSelectable(spec, {})).toMatchSnapshot();
  });
});
