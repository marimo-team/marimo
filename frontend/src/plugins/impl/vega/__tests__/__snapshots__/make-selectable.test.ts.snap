// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`makeSelectable > should return correctly if existing legend selection 1`] = `
{
  "config": {
    "view": {
      "continuousHeight": 300,
    },
  },
  "data": {
    "url": "https://cdn.jsdelivr.net/npm/vega-datasets@v1.29.0/data/unemployment-across-industries.json",
  },
  "encoding": {
    "color": {
      "field": "series",
      "scale": {
        "scheme": "category20b",
      },
      "type": "nominal",
    },
    "opacity": {
      "condition": {
        "test": {
          "and": [
            {
              "param": "param_1",
            },
            {
              "param": "legend_selection_series",
            },
            {
              "param": "select_point",
            },
          ],
        },
        "value": 1,
      },
      "value": 0.2,
    },
    "x": {
      "axis": {
        "domain": false,
        "format": "%Y",
        "tickSize": 0,
      },
      "field": "date",
      "timeUnit": "yearmonth",
      "type": "temporal",
    },
    "y": {
      "aggregate": "sum",
      "axis": null,
      "field": "count",
      "stack": "center",
      "type": "quantitative",
    },
  },
  "mark": {
    "cursor": "pointer",
    "tooltip": true,
    "type": "area",
  },
  "params": [
    {
      "bind": "legend",
      "name": "param_1",
      "select": {
        "fields": [
          "series",
        ],
        "type": "point",
      },
    },
    {
      "bind": "legend",
      "name": "legend_selection_series",
      "select": {
        "fields": [
          "series",
        ],
        "type": "point",
      },
    },
    {
      "name": "select_point",
      "select": {
        "encodings": [
          "color",
        ],
        "type": "point",
      },
    },
  ],
}
`;

exports[`makeSelectable > should return correctly if mark is a string 1`] = `
{
  "encoding": {
    "opacity": {
      "condition": {
        "test": {
          "and": [
            {
              "param": "select_point",
            },
            {
              "param": "select_interval",
            },
          ],
        },
        "value": 1,
      },
      "value": 0.2,
    },
  },
  "mark": {
    "cursor": "pointer",
    "tooltip": true,
    "type": "point",
  },
  "params": [
    {
      "name": "select_point",
      "select": {
        "encodings": [
          "x",
          "y",
        ],
        "type": "point",
      },
    },
    {
      "name": "select_interval",
      "select": {
        "encodings": [
          "x",
          "y",
        ],
        "mark": {
          "fill": "#669EFF",
          "fillOpacity": 0.07,
          "stroke": "#669EFF",
          "strokeOpacity": 0.4,
        },
        "type": "interval",
      },
    },
  ],
}
`;

exports[`makeSelectable > should return correctly if mark is not string 1`] = `
{
  "encoding": {
    "opacity": {
      "condition": {
        "test": {
          "and": [
            {
              "param": "select_point",
            },
            {
              "param": "select_interval",
            },
          ],
        },
        "value": 1,
      },
      "value": 0.2,
    },
  },
  "mark": {
    "cursor": "pointer",
    "tooltip": true,
    "type": "point",
  },
  "params": [
    {
      "name": "select_point",
      "select": {
        "encodings": [
          "x",
          "y",
        ],
        "type": "point",
      },
    },
    {
      "name": "select_interval",
      "select": {
        "encodings": [
          "x",
          "y",
        ],
        "mark": {
          "fill": "#669EFF",
          "fillOpacity": 0.07,
          "stroke": "#669EFF",
          "strokeOpacity": 0.4,
        },
        "type": "interval",
      },
    },
  ],
}
`;

exports[`makeSelectable > should return correctly if overlapping encodings 1`] = `
{
  "config": {
    "view": {
      "continuousHeight": 300,
    },
  },
  "data": {
    "url": "data/cars.json",
  },
  "encoding": {
    "color": {
      "field": "Origin",
      "type": "nominal",
    },
    "opacity": {
      "condition": {
        "test": {
          "and": [
            {
              "param": "legend_selection_Origin",
            },
            {
              "param": "select_point",
            },
            {
              "param": "select_interval",
            },
          ],
        },
        "value": 1,
      },
      "value": 0.2,
    },
    "x": {
      "field": "Horsepower",
      "type": "quantitative",
    },
    "y": {
      "field": "Miles_per_Gallon",
      "type": "quantitative",
    },
  },
  "mark": {
    "cursor": "pointer",
    "tooltip": true,
    "type": "point",
  },
  "params": [
    {
      "bind": "legend",
      "name": "legend_selection_Origin",
      "select": {
        "fields": [
          "Origin",
        ],
        "type": "point",
      },
    },
    {
      "name": "select_point",
      "select": {
        "encodings": [
          "x",
          "y",
        ],
        "type": "point",
      },
    },
    {
      "name": "select_interval",
      "select": {
        "encodings": [
          "x",
          "y",
        ],
        "mark": {
          "fill": "#669EFF",
          "fillOpacity": 0.07,
          "stroke": "#669EFF",
          "strokeOpacity": 0.4,
        },
        "type": "interval",
      },
    },
  ],
}
`;

exports[`makeSelectable > should return correctly with multiple encodings 1`] = `
{
  "encoding": {
    "color": {
      "field": "colorField",
      "type": "nominal",
    },
    "opacity": {
      "condition": {
        "test": {
          "and": [
            {
              "param": "legend_selection_colorField",
            },
            {
              "param": "legend_selection_sizeField",
            },
            {
              "param": "select_point",
            },
            {
              "param": "select_interval",
            },
          ],
        },
        "value": 1,
      },
      "value": 0.2,
    },
    "size": {
      "field": "sizeField",
      "type": "quantitative",
    },
  },
  "mark": {
    "cursor": "pointer",
    "tooltip": true,
    "type": "point",
  },
  "params": [
    {
      "bind": "legend",
      "name": "legend_selection_colorField",
      "select": {
        "fields": [
          "colorField",
        ],
        "type": "point",
      },
    },
    {
      "bind": "legend",
      "name": "legend_selection_sizeField",
      "select": {
        "fields": [
          "sizeField",
        ],
        "type": "point",
      },
    },
    {
      "name": "select_point",
      "select": {
        "encodings": [
          "x",
          "y",
        ],
        "type": "point",
      },
    },
    {
      "name": "select_interval",
      "select": {
        "encodings": [
          "x",
          "y",
        ],
        "mark": {
          "fill": "#669EFF",
          "fillOpacity": 0.07,
          "stroke": "#669EFF",
          "strokeOpacity": 0.4,
        },
        "type": "interval",
      },
    },
  ],
}
`;

exports[`makeSelectable > should skip field selection if empty or false 1`] = `
{
  "config": {
    "view": {
      "continuousHeight": 300,
    },
  },
  "data": {
    "url": "data/cars.json",
  },
  "encoding": {
    "color": {
      "field": "Origin",
      "type": "nominal",
    },
    "opacity": {
      "condition": {
        "test": {
          "and": [
            {
              "param": "select_point",
            },
            {
              "param": "select_interval",
            },
          ],
        },
        "value": 1,
      },
      "value": 0.2,
    },
    "x": {
      "field": "Horsepower",
      "type": "quantitative",
    },
    "y": {
      "field": "Miles_per_Gallon",
      "type": "quantitative",
    },
  },
  "mark": {
    "cursor": "pointer",
    "tooltip": true,
    "type": "point",
  },
  "params": [
    {
      "name": "select_point",
      "select": {
        "encodings": [
          "x",
          "y",
        ],
        "type": "point",
      },
    },
    {
      "name": "select_interval",
      "select": {
        "encodings": [
          "x",
          "y",
        ],
        "mark": {
          "fill": "#669EFF",
          "fillOpacity": 0.07,
          "stroke": "#669EFF",
          "strokeOpacity": 0.4,
        },
        "type": "interval",
      },
    },
  ],
}
`;

exports[`makeSelectable > should work for multi-layered charts 1`] = `
{
  "data": {
    "name": "source",
  },
  "datasets": {
    "source": [
      {
        "yield_center": 32.4,
        "yield_error": 7.5522,
      },
      {
        "yield_center": 30.96667,
        "yield_error": 6.9775,
      },
      {
        "yield_center": 33.966665,
        "yield_error": 3.9167,
      },
      {
        "yield_center": 30.45,
        "yield_error": 11.9732,
      },
    ],
  },
  "layer": [
    {
      "encoding": {
        "x": {
          "field": "yield_center",
          "scale": {
            "zero": false,
          },
          "title": "yield",
          "type": "quantitative",
        },
        "xError": {
          "field": "yield_error",
        },
        "y": {
          "field": "variety",
          "type": "nominal",
        },
      },
      "mark": {
        "ticks": true,
        "type": "errorbar",
      },
    },
    {
      "encoding": {
        "opacity": {
          "condition": {
            "test": {
              "and": [
                {
                  "param": "select_point",
                },
                {
                  "param": "select_interval",
                },
              ],
            },
            "value": 1,
          },
          "value": 0.2,
        },
        "x": {
          "field": "yield_center",
          "type": "quantitative",
        },
      },
      "mark": {
        "color": "black",
        "cursor": "pointer",
        "filled": true,
        "tooltip": true,
        "type": "point",
      },
      "params": [
        {
          "name": "select_point",
          "select": {
            "encodings": [
              "x",
              "y",
            ],
            "type": "point",
          },
        },
        {
          "name": "select_interval",
          "select": {
            "encodings": [
              "x",
              "y",
            ],
            "mark": {
              "fill": "#669EFF",
              "fillOpacity": 0.07,
              "stroke": "#669EFF",
              "strokeOpacity": 0.4,
            },
            "type": "interval",
          },
        },
      ],
    },
  ],
  "width": "container",
}
`;
